viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -526,31 +526,40 @@ void SLAMServiceImpl::RunSLAM() {
 }
 
 std::string SLAMServiceImpl::GetNextDataFileOffline() {
-    if (!b_continue_session) {
-        return "";
-    }
-    if (file_list_offline.size() == 0) {
-        file_list_offline = viam::io::ListSortedFilesInDirectory(path_to_data);
-    }
-    // We're setting the minimum required files to be two for the following
-    // reasons:
-    // 1. Cartographer needs at least two PCD files to work properly.
-    // 2. A .DS_Store file is frequently added to the data directory when
-    // a user opens the directory on osx.
-    // Expecting a minimum of 3 files solves both problems without having to
-    // loop over and count the number of actual data files in the data
-    // directory.
-    if (file_list_offline.size() <= 2) {
-        throw std::runtime_error("not enough data in data directory");
-    }
-    if (current_file_offline == file_list_offline.size()) {
+    while (b_continue_session) {
+        if (file_list_offline.size() - current_file_offline <= 2) {
+            file_list_offline =
+                viam::io::ListSortedFilesInDirectory(path_to_data);
+        }
+        // We're setting the minimum required files to be two for the following
+        // reasons:
+        // 1. Cartographer needs at least two PCD files to work properly.
+        // 2. A .DS_Store file is frequently added to the data directory when
+        // a user opens the directory on osx.
+        // Expecting a minimum of 3 files solves both problems without having to
+        // loop over and count the number of actual data files in the data
+        // directory.
+        if (file_list_offline.size() > 2) {
+            // If the current file is the last in the file list, skip it, since
+            // it's possible it's still being written to.
+            if (current_file_offline != file_list_offline.size() - 1) {
+                const auto to_return = file_list_offline[current_file_offline];
+                if (to_return != "") {
```

Comment:
This might not be necessary anymore actually; the last file in the list was showing up as "", not sure if that's a C++ null termination thing or a shortcoming of how our file list is being constructed.